### PR TITLE
[vulkan] Isolate SPIR-V codegen, cleanup Vulkan backend

### PR DESCRIPTION
**Pull request by bobcao3**
Related issue = #

### CODE REVIEW

File: taichi/codegen/spirv/spirv_codegen.cpp [hunk 1]
```diff
@@ -1738,96 +1738,69 @@ static void spriv_message_consumer(spv_message_level_t level,
...
-class KernelCodegen {
- public:
-  struct Params {
-    std::string ti_kernel_name;
-    Kernel *kernel;
-    std::vector<CompiledSNodeStructs> compiled_structs;
-    Device *device;
-    bool enable_spv_opt{true};
-  };
+static std::unique_ptr<spvtools::Optimizer> spirv_opt_{nullptr};
```
[the remark this answers is not in the document]
Comment (bobcao3) [reply taichi/codegen/spirv/spirv_codegen.cpp:1741]: Should I put it as a static variable in like an get_instance function or something else?

[the remark this answers is not in the document]
Comment (k-ye) [reply taichi/codegen/spirv/spirv_codegen.cpp:1741]: I think it's fine to just place it in the codegen itself. Was the original intention to save a few rounds of spv optimizer's allocation?

[the remark this answers is not in the document]
Comment (bobcao3) [reply taichi/codegen/spirv/spirv_codegen.cpp:1741]: ^ yes

[the remark this answers is not in the document]
Comment (k-ye) [reply taichi/codegen/spirv/spirv_codegen.cpp:1741]: Cool, let's just put these into `KernelCodegen`

File: taichi/codegen/spirv/spirv_codegen.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  void run(TaichiKernelAttributes &kernel_attribs,
+           std::vector<std::vector<uint32_t>> &generated_spirv);
+
+ private:
+  Params params_;
+  KernelContextAttributes ctx_attribs_;
+
+  std::unique_ptr<spvtools::Optimizer> spirv_opt_{nullptr};
+  std::unique_ptr<spvtools::SpirvTools> spirv_tools_{nullptr};
+  spvtools::OptimizerOptions _spirv_opt_options;
```
Comment (k-ye) [inline taichi/codegen/spirv/spirv_codegen.h:41]:
```suggestion
  spvtools::OptimizerOptions spirv_opt_options_;
```

**Review by k-ye — APPROVED** [review]
LGTM!

**Pull request merged**